gammapy/utils/types.py

```diff
@@ -0,0 +1,128 @@
+from pathlib import Path
+from typing import Annotated
+from astropy.coordinates import Angle, EarthLocation, SkyCoord
+from astropy.time import Time
+from astropy.units import Quantity
+from pydantic import PlainSerializer
+from pydantic.functional_validators import BeforeValidator
+from .observers import observatory_locations
+from .scripts import make_path
+
+__all__ = [
+    "AngleType",
+    "EnergyType",
+    "TimeType",
+    "PathType",
+    "EarthLocationType",
+    "SkyCoordType",
+]
+
+
+def json_encode_quantity(v):
+    """JSON encoder for `~astropy.units.Quantity`."""
+    return f"{v.value} {v.unit}"
+
+
+def json_encode_angle(v):
+    """JSON encoder for `~astropy.coordinates.Angle`."""
+    return f"{v.value} {v.unit}"
+
+
+def json_encode_time(v):
+    """JSON encoder for `~astropy.time.Time`."""
+    return f"{v.value}"
+
+
+def json_encode_earth_location(v):
+    """JSON encoder for `~astropy.coordinates.EarthLocation`."""
+    return (
+        f"lon: {v.lon.value} {v.lon.unit}, "
+        f"lat : {v.lat.value} {v.lat.unit}, "
+        f"height : {v.height.value} {v.height.unit}"
+    )
+
+
+def json_encode_sky_coord(v):
+    """JSON encoder for `~astropy.coordinates.SkyCoord`."""
+    return f"lon: {v.spherical.lon.value} {v.spherical.lon.unit}, lat: {v.spherical.lat.value} {v.spherical.lat.unit}, frame: {v.frame.name} "
+
+
+def validate_angle(v):
+    """Validator for `~astropy.coordinates.Angle`."""
+    return Angle(v)
+
+
+def validate_energy(v):
+    """Validator for `~astropy.units.Quantity` with unit "energy"."""
+    v = Quantity(v)
+    if v.unit.physical_type != "energy":
+        raise ValueError(f"Invalid unit for energy: {v.unit!r}")
+    return v
+
+
+def validate_time(v):
+    """Validator for `~astropy.time.Time`."""
+    return Time(v)
+
+
+def validate_earth_location(v):
+    """Validator for `~astropy.coordinates.EarthLocation`."""
+    if isinstance(v, EarthLocation):
+        return v
+
+    if isinstance(v, str) and v in observatory_locations:
+        return observatory_locations[v]
+
+    try:
+        return EarthLocation(v)
+    except TypeError:
+        raise ValueError(f"Invalid EarthLocation: {v!r}")
+
+
+def validate_sky_coord(v):
+    """Validator for `~astropy.coordinates.SkyCoord`."""
+    return SkyCoord(v)
+
+
+AngleType = Annotated[
+    Angle,
+    PlainSerializer(json_encode_angle, return_type=str, when_used="json-unless-none"),
+    BeforeValidator(validate_angle),
+]
+
+EnergyType = Annotated[
```

Comment:
Thanks, I was not aware of this. However in our case, we mostly do it for the custom validation and serialization methods and only the energy type is handled with a quantity object. I guess Pydantic does not understand the `PhysicalType` annotation, so we need a validator in any case.

Reply:
This seems to work nicely:

```
from typing import Annotated, get_args
from pydantic import BaseModel, BeforeValidator, ConfigDict, ValidationError
from pydantic.functional_validators import AfterValidator
import astropy.units as u
from functools import partial


def validate_physical_type(value: u.Quantity, physical_type: u.PhysicalType):
    if value.unit is None:
        raise ValueError(f"Expected physical_type: {physical_type}, but value had no unit")

    if (t := value.unit.physical_type) != physical_type:
        raise ValueError(f"Expected physical_type: {physical_type}, got {t}")

    return value


class ValidatedQuantity(u.Quantity):
    @classmethod
    def __class_getitem__(cls, physical_type: str):
        q = super().__class_getitem__(physical_type)
        validate = partial(validate_physical_type, physical_type=get_args(q)[1])
        return Annotated[
            q,
            BeforeValidator(lambda v: cls(v, copy=False)),
            AfterValidator(validate)
        ]


class Foo(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)
    length : ValidatedQuantity['length'] = 5 * u.m


Foo(length=5 * u.cm)

try:
    Foo(length=5 * u.deg)
except ValidationError as e:
    print(e)

```

Reply:
Thanks this looks generally useful! I think for this PR it does not add any value, because currently we only have the `EnergyType`. But I'm pretty sure we will need it in future, so maybe you just want to contribute it in a follow up PR yourself?